x-pack/plugins/lists/common/types.ts

```diff
@@ -0,0 +1,30 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+/**
+ * This makes any optional property the same as Required<T> would but also has the
+ * added benefit of keeping your undefined.
+ *
+ * For example:
+ * type A = RequiredKeepUndefined<{ a?: undefined; b: number }>;
+ *
+ * will yield a type of:
+ * type A = { a: undefined; b: number; }
+ *
+ */
+export type RequiredKeepUndefined<T> = { [K in keyof T]-?: [T[K]] } extends infer U
+  ? U extends Record<keyof U, [unknown]>
+    ? { [K in keyof U]: U[K][0] }
+    : never
+  : never;
```

Comment:
NOTE: See docs above, this will remove the `?` but keep the `undefined` when used with io-ts so that we can force push down `undefined` outside of the validation within the code base. This makes it easier for functions to have to be required to deal with types rather than ignore them by accident.

Reply:
Actually just saw that you're already using unions for those values! So we could still surface let's say `id` and `meta` here as optional (as to whether they're even passed in) using the type defined by `Identity` but then use the type defined by `RequiredKeepUndefined` throughout our code to always explicitly deal with the `undefined`s?

Reply:
I think I am doing that if I understand what you're saying correctly.

I don't use the partials anywhere within the lists code base. I only added the partials for front end code to potentially use and any tests that want to test out absent keys and values to ensure that the REST requests work out correctly.

I do want to force all of the code under the covers to _always_ deal with the key and value of the union of `value or undefined` even if it has to be explicit in passing in an `undefined`.

The partials are only there for the front end to use. If the front end for some reason never uses them and always pushes `undefined` as the value then I will potentially remove the extra type.

The interesting thing about:

```ts
JSON.stringify({ key: undefined });
```

Is that it returns:

```ts
{}
```

But I don't want to force the front end to have to populate `undefined` if it makes more sense for it just remove keys. But the front end can decide if it wants to use the partials if that makes more sense or if it wants to force the presence of undefined or if it wants to use a mix between the two.

Reply:
Yea, you worded it much better, but that's what I was wondering if you were intending to use these as :)